lib/galaxy/jobs/__init__.py

```diff
@@ -1591,7 +1591,15 @@ def fail():
 
         outputs_to_working_directory = util.asbool(self.get_destination_configuration("outputs_to_working_directory", False))
         if outputs_to_working_directory and not self.__link_file_check():
-            for dataset_path in self.get_output_fnames():
+            if self.output_paths is None:
+                self.compute_outputs()
+            for (name, (_, dataset_path)) in self.output_hdas_and_paths.items():
+                if dataset_path.false_path and not os.path.exists(dataset_path.false_path):
+                    output = self.tool.find_output_def(name)
+                    if output and getattr(output, 'from_work_dir') is not None:
+                        # This doesn't actually have to be present, unfortunately we cannot distinguish between
+                        # NFS issues and missing outputs in this case.
```

Comment:
This fails with external metadata setting for collection_creates_pair_from_work_dir, but not output_empty_work_dir, probably because output_empty_work_dir doesn't define a datatype:
```
galaxy.metadata DEBUG 2019-11-04 16:03:05,017 setting metadata externally failed for HistoryDatasetAssociation 114: Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/metadata/set_metadata.py", line 137, in set_metadata_portable
    set_meta(dataset, file_dict)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/metadata/set_metadata.py", line 107, in set_meta
    set_meta_with_tool_provided(new_dataset_instance, file_dict, set_meta_kwds, datatypes_registry, max_metadata_value_size)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/metadata/set_metadata.py", line 68, in set_meta_with_tool_provided
    dataset_instance.datatype.set_meta(dataset_instance, **set_meta_kwds)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/datatypes/data.py", line 816, in set_meta
    dataset.metadata.data_lines = self.count_data_lines(dataset)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/datatypes/data.py", line 835, in count_data_lines
    with compression_utils.get_fileobj(dataset.file_name) as in_file:
  File "/Users/mvandenb/src/galaxy/lib/galaxy/util/compression_utils.py", line 30, in get_fileobj
    return get_fileobj_raw(filename, mode, compressed_formats)[1]
  File "/Users/mvandenb/src/galaxy/lib/galaxy/util/compression_utils.py", line 61, in get_fileobj_raw
    return compressed_format, io.open(filename, mode, encoding='utf-8')
FileNotFoundError: [Errno 2] No such file or directory: '/Users/mvandenb/src/galaxy/database/jobs_directory/000/109/outputs/galaxy_dataset_114.dat'
```

I'm not sure this is a good idea in general, it feels like failing is the proper option here given that nothing indicates that `from_work_dir` will work even if the file doesn't exist. Even if we'd fall back and just `touch` the output file setting the metadata for some datatypes is going to fail (can't `pysam.index()` and empty file for instance) -- and that's also going to fail now if `outputs_to_working_directory` is disabled.

We could introduce a new flag for optional outputs though. It'd still have the metadata problems, but at least it's explicitly allowed to not be present by the tool author. And you could also just discover the datasets if uncertain, or `touch` all `from_work_dir` outputs in the wrapper.

Reply:
I think from a backwards compatibility perspective, the proposed flag might be backwards. Since `outputs_to_working_directory` defaults to `False`, assuming the behavior of `from_work_dir` had been Green/OK (haven't actually checked older checkouts), existing behavior should be maintained.

Regardless, things of this nature at a tool-design level need to be handled consistently regardless of the job configurations.